Extend `Print Canonical Projections` with a search functionality

pi8027 opened the pull request:
To see how canonical projections work and which projection/pattern they involve in advanced canonical structure programming, we need to use the `Print Canonical Projections` command and find a unification rule from a bunch of output lines. This PR allows `Print Canonical Projections` to take constants and print only the unification rules that involve or are synthesized from given constants.

<!-- Keep what applies -->
**Kind:** feature.

<!-- If there is a user-visible change in coqc/coqtop/coqchk/coq_makefile behavior and testing is not prohibitively expensive: -->
<!-- (Otherwise, remove this line.) -->
- [x] Added / updated test-suite
<!-- If this is a feature pull request / breaks compatibility: -->
<!-- (Otherwise, remove these lines.) -->
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

ejgallego commented on pretyping/recordops.ml:
```diff
@@ -246,13 +246,23 @@ let warn_redundant_canonical_projection =
+
+let print_canonical_projections env sigma cs = !canonical_projections_printer env sigma cs
+
 let register_canonical_structure ~warn env sigma o =
```
Indeed that's the wrong place to add this print statement.

pi8027 replied:
Yes. I didn't expect this behavior when I start this.

pi8027 commented:
@ejgallego So could we have a new option to control this verbosity, or change the `About` command to print canonical projections associated with the given constant? (But `object_table` in `recordops.ml` isn't designed in this way so now I don't see how to do it.)

SkySkimmer commented:
We can always add a table if it's useful

Gaëtan Gilbert

On 11/09/2019 15:32, Kazuhiko Sakaguchi wrote:
> @ejgallego <https://github.com/ejgallego> So could we have a new option 
> to control this verbosity, or change the |About| command to print 
> canonical projections associated with the given constant? (But 
> |object_table| in |recordops.ml| isn't designed in this way so now I 
> don't see how to do it.)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coq/coq/pull/10747?email_source=notifications&email_token=AASZB3AXBKSXMD7Q3DLZM6LQJDXQHA5CNFSM4IVO5EM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6OPZWY#issuecomment-530382043>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AASZB3D7TO73R32JNE22AV3QJDXQHANCNFSM4IVO5EMQ>.
>

ejgallego commented:
Yeah improving the table/`About` command seems right.

pi8027 commented:
@SkySkimmer @ejgallego I will do so. Thanks!

pi8027 commented:
I have almost finished the implementation, but I'm still not sure about my design. I will discuss with someone to make sure of it. (Maybe I need to change `Print Canonical Projection` to take a constant optionally rather than changing the `About` command.)

pi8027 commented:
Since I have no idea on how this should work with PR #10984 (declaring multiple instances at once), I would like to have comments from @gares and @CohenCyril.

gares commented:
If you want to scale this to the other pr you should make the other pr keep a link between what the user wrote and the derived cs instances I'm afraid.

pi8027 commented:
So IIUC, I should extend the object table ‘obj_typ’ with a gref. Using ‘o_DEF’ for this purpose seems to be a bad idea anyway.

gares commented:
Maybe o_ORIGIN : gref option. But you should not wait for the other pr, this change morally belongs to the other one.

pi8027 commented:
@gares Since I didn't see why you prefer `option`, I put `o_ORIGIN : Constant.t`.

gares commented:
It is OK.

I was proposing option because for regular CS you have o_DEF = o_ORIGIN, and you have != for multiple CS. With option you could have None for the first case, and (Some origin) for the second.
Your code tests for equality, while it could match on the option... not very different.

herbelin commented:
Looks relevant and very good to me. (But I must confess that I did not understand @ejgallego's objection about noisiness. Isn't the PR only allowing to reduce the verbosity of the `Print Canonical Projections` query?)

ejgallego commented:
> But I must confess that I did not understand @ejgallego's objection about noisiness. Isn't the PR only allowing to reduce the verbosity of the `Print Canonical Projections` query?

My comment was referring to the original proposal of print the projections after each `Canonical` commands, I think the current proposal doesn't do that so my review can be disregarded.

pi8027 commented:
@ejgallego @herbelin Yes. I changed the design of this feature twice. Old implementations were discarded but they can be seen in the changes of the title.

I have just added a comment and rebased again. I think this is ready to merge if CI succeeds.

CohenCyril commented on doc/sphinx/language/gallina-extensions.rst:
```diff
@@ -2075,11 +2075,13 @@ in :ref:`canonicalstructures`; here only a simple example is given.
-   which it is a projection is indicated.
+   which it is a projection is indicated. If constants are given as
+   its arguments, only the unification rules that involve or are
+   synthesized from given constants will be shown.
```
I suggest the following
```suggestion
   synthesized from simultaneously all given constants will be shown.
```
to disambiguate that all constraints must be satisfied and not at least one.

gares commented:
@pi8027 if you rebase and implement the fix requested by Cyril I'll happily merge (CI failure is unrelated)

pi8027 commented:
@gares Done.
